# Course Content


| Days     | Session                                                  |
|----------|----------------------------------------------------------|
| 1-2      | Python Basics                                            |
| 3        | Classes in Python and Make a Weather App                 |
| 4        | Hangman Game                                             |
| 5        | Numpys and Pandas                                        |
| 6        | Matplot lib and Seaborn                                  |
| 7        | Exploratory Data Analysis                                |
| 8,9,10   | End to End Data Analysis Project (With local deployment) |
| 11,12,13 | End to End Data Analysis Project (With local deployment) |
| 14       | Discussion,Complete Roadmap on Data Science,ML and AI    |


# What is AI? ML? DL? DS?

It is the 21st century, and technology is changing faster than ever. To meet the market current opportunities, we should know Artificial Intelligence vs Machine Learning vs Deep Learning vs Data Science.

While the terms Data Science, Artificial Intelligence (AI) and Machine learning fall in the same domain and are connected to each other, they have their specific applications and meaning. There may be overlaps in these domains every now and then, but essentially, each of these four terms has unique uses of its own. 


So, are you ready to explore AI vs ML vs DL vs DS?





![](https://d2h0cx97tjks2p.cloudfront.net/blogs/wp-content/uploads/sites/2/2019/02/Ai-vs-ml-vs-dl-vs-ds.jpg)

## Artificial Intelligence

![](https://upload.wikimedia.org/wikipedia/commons/2/2e/What-is-artificial-intelligence-_What_is_artificial_intelligence.jpg)



*Artificial Intelligence, fondly abbreviated as AI, is concerned with imparting human intelligence to machines. It focuses on the development of intelligent machines that can think and act like humans; essentially, AI is intelligence such as machines display.*

An intelligent agent is a device that can perceive its environment and act to optimize its chances of success. Such intelligent machines mimic human cognitive functions like learning and problem-solving.

Artificial Intelligence is the intelligence that machines can portray- they can think and act like humans. They can perform logical reasoning, learning, and also self-correction. AI can make predictions and take decisions.

AI deals with the following issues-

1. Reasoning and Problem Solving
2. Knowledge representation
3. Planning
4. Learning
5. Natural Language Processing (NLP)
6. Perception
7. Motion and Manipulation
8. Social Intelligence
9. General Intelligence


Types:

* General AI
* Narrow AI

While general AI is machine intelligence that can perform any intellectual task that a human can, narrow AI often does specific tasks and does it better than humans. While general AI is still a dream to achieve, the AlphaGo can be an example of narrow AI.




AlphaGo versus Lee Sedol, also known as the Google DeepMind Challenge Match, was a five-game Go match between 18-time world champion Lee Sedol and AlphaGo, a computer Go program developed by Google DeepMind, played in Seoul, South Korea between 9 and 15 March 2016. AlphaGo won all but the fourth game; all games were won by resignation.


![](https://s.abcnews.com/images/Technology/HT_Google_AI_Go_hb_160309_16x9_992.jpg)





## Machine Learning


![](https://dlaignite.com/wp-content/uploads/2018/02/machine-learning.png)

In simple terms machine learning is a sub field of artificial intelligence in which we study algorithms to make prediction based on training data that improve automatically through experiences. 

Machine Learning is concerned with giving machines the ability to learn by training algorithms on a huge amount of data. It makes use of algorithms and statistical models to perform a task without needing explicit instructions.


There are three types of learning here:

1. Supervised and semi-supervised learning
2. Unsupervised learning
3. Reinforcement learning

![](https://miro.medium.com/max/2000/1*8wU0hfUY3UK_D8Y7tbIyFQ.png)



### SUPERVISED LEARNING



![](https://sebastianraschka.com/images/blog/2014/intro_supervised_learning/learning_algorithm_1.png)


### ML examples

For instance, if you show one such algorithm enough pictures of a puppy, it will identify one in a picture it has never seen before. Try it with pictures of Chihuahuas and those of muffins, or with pictures of white Pulis and those of mops.

![](https://d2h0cx97tjks2p.cloudfront.net/blogs/wp-content/uploads/sites/2/2019/02/Image-recognition-with-ML.jpg)

![](https://d2h0cx97tjks2p.cloudfront.net/blogs/wp-content/uploads/sites/2/2019/02/Image-recognition-with-ML.jpg)


Another interesting example is the website [GANS](https://thispersondoesnotexist.com/). It uses the ML algorithm Generative Adversarial Network (GAN) to create images of people that do not exist; it displays a new picture in 1024×1024 every time you refresh the website, and it does so in High-res.

### Then how AI and ML are different?


Well, ML is an approach to AI (it is a way of achieving Artificial Intelligence). It is possible to achieve Artificial Intelligence without ML, but that can take millions of lines of code.



## Deep Learning

Deep Learning is an approach to Machine Learning; one that focuses on learning data representations rather than on task-specific algorithms. It makes use of Deep Neural Networks, which are inspired by the structure and function of the human brain.

### DL algorithms

* Convolutional Neural Networks (CNNs) 
* Recurrent Neural Networks (RNNs)
* Generative Adversarial Networks (GANs) 

![](https://d2h0cx97tjks2p.cloudfront.net/blogs/wp-content/uploads/sites/2/2019/02/Deep-learning-introduction.png)


Deep Learning is used for purposes like Natural Language Processing (NLP),Image Processing.








## Data Science and Data Scientist


Data science is a multidisciplinary approach to extracting actionable insights from the large and ever-increasing volumes of data collected and created by today’s organizations.

Data science encompasses preparing data for analysis and processing, performing advanced data analysis, and presenting the results to reveal patterns and enable stakeholders to draw informed conclusions.

Data preparation can involve cleansing, aggregating, and manipulating it

So, basically in simple terms data science refers knowledge of data. 


**Data Science:**  Data science is basically a process of using data by extracting, cleaning and manipulating data to make a decision or to predict a outcome using cleaned data. 

**Data Scientist:** data science practitioners

**Data science tools:** 

* R
* Python




CONCLUSION:

![](https://d2h0cx97tjks2p.cloudfront.net/blogs/wp-content/uploads/sites/2/2019/02/Artificial-Intelligence-vs-Machine-Learning-vs-Deep-Learning-01.jpg)

# Why Python



1.   Open Source
2.   Large Community
3. Countless Libraries for Data Science
4. Countless Machine Learning Libraries
5. Relatively simple Syntax (easy for beginners to learn to code)
6. used for web development, scientific/numeric computing, game development




# First Steps with Python and Jupyter 

![](https://i.imgur.com/gvSnw4A.png)




This tutorial covers the following topics:

* Performing arithmetic operations using Python
* Solving multi-step problems using variables
* Evaluating conditions using Python
* Combining conditions with logical operators
* Adding text styles using Markdown

## Performing Arithmetic Operations using Python

Let's begin by using Python as a calculator. You can write and execute Python using a code cell within Jupyter. 

> **Working with cells**: To create a new cell within Jupyter, you can select "Insert > Insert Cell Below" from the menu bar or just press the "+" button on the toolbar. You can also use the keyboard shortcut `Esc+B` to create a new cell. Once a cell is created, click on it to select it. You can then change the cell type to code or markdown (text) using the "Cell > Cell Type" menu option. You can also use the keyboard shortcuts `Esc+Y` and `Esc+M`. Double-click a cell to edit the content within the cell. To apply your changes and run a cell, use the "Cell > Run Cells" menu option or click the "Run" button on the toolbar or just use the keyboard shortcut `Shift+Enter`. You can see a full list of keyboard shortcuts using the "Help > Keyboard Shortcuts" menu option.

Run the code cells below to perform calculations and view their result. Try changing the numbers and run the modified cells again to see updated results. Can you guess what the `//`, `%`, and `**` operators are used for?

In [5]:
print(2 + 3 + 9)

14


In [3]:
print(type(1))

<class 'int'>


In [6]:
99 - 73

26

In [8]:
23.54 * -1432

-33709.28

In [9]:
100 / 7

14.285714285714286

In [10]:
100 // 7

14

In [11]:
100 % 7

2

In [12]:
10 % 3

1

In [13]:
5 * 3

15

In [14]:
5 * 5 * 5 

125

In [15]:
5 ** 3

125

As you might expect, operators like `/` and `*` take precedence over other operators like `+` and `-` as per mathematical conventions. You can use parentheses, i.e. `(` and `)`, to specify the order in which operations are performed.

In [16]:
((2 + 5) * (17 - 3)) / (4 ** 3)

1.53125

Python supports the following arithmetic operators:

| Operator   | Purpose           | Example     | Result    |
|------------|-------------------|-------------|-----------|
| `+`        | Addition          | `2 + 3`     | `5`       |
| `-`        | Subtraction       | `3 - 2`     | `1`       |
| `*`        | Multiplication    | `8 * 12`    | `96`      |
| `/`        | Division          | `100 / 7`   | `14.28..` |
| `//`       | Floor Division    | `100 // 7`  | `14`      |    
| `%`        | Modulus/Remainder | `100 % 7`   | `2`       |
| `**`       | Exponent          | `5 ** 3`    | `125`     |


Try solving some simple problems from this page:
https://www.math-only-math.com/worksheet-on-word-problems-on-four-operations.html . 

You can use the empty cells below and add more cells if required.

## Solving multi-step problems using variables

Let's try solving the following word problem using Python: 

> A grocery store sells a bag of ice for $1.25 and makes a 20% profit. If it sells 500 bags of ice, how much total profit does it make?

We can list out the information provided and gradually convert the word problem into a mathematical expression that can be evaluated using Python. 

*Cost of ice bag ($)* = 1.25

*Profit margin* = 20% = .2

*Profit per bag ($)* = profit margin * cost of ice bag = .2 * 1.25

*No. of bags* = 500

*Total profit* = no. of bags * profit per bag = 500 * (.2 * 1.25)

In [17]:
500 * (.2 * 1.25)

125.0

Thus, the grocery store makes a total profit of $125. While this is a reasonable way to solve a problem, it's not entirely clear by looking at the code cell what the numbers represent. We can give names to each of the numbers by creating Python *variables*.

> **Variables**: While working with a programming language such as Python, information is stored in *variables*. You can think of variables as containers for storing data. The data stored within a variable is called its *value*.

In [18]:
cost_of_ice_bag = 1.25

In [19]:
profit_margin = .2

In [20]:
number_of_bags = 500

The variables `cost_of_ice_bag`, `profit_margin`, and `number_of_bags` now contain the information provided in the word problem. We can check the value of a variable by typing its name into a cell. We can combine variables using arithmetic operations to create other variables.

> **Code completion**: While typing the name of an existing variable in a code cell within Jupyter, just type the first few characters and press the `Tab` key to autocomplete the variable's name. Try typing `pro` in a code cell below and press `Tab` to autocomplete to `profit_margin`.

In [21]:
profit_margin

0.2

In [22]:
profit_per_bag = cost_of_ice_bag * profit_margin

In [23]:
profit_per_bag

0.25

In [24]:
total_profit = number_of_bags * profit_per_bag

In [25]:
total_profit

125.0

If you try to view the value of a variable that has not been *defined*, i.e., given a value using the assignment statement `variable_name = value`, Python shows an error.

In [26]:
a

NameError: ignored

Storing and manipulating data using appropriately named variables is a great way to explain what your code does.

Let's display the result of the word problem using a friendly message. We can do this using the `print` *function*.

> **Functions**: A function is a reusable set of instructions. It takes one or more inputs, performs certain operations, and often returns an output. Python provides many in-built functions like `print` and also allows us to define our own functions.

In [27]:
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 125.0


> **`print`**: The `print` function is used to display information. It takes one or more inputs, which can be text (within quotes, e.g., `"this is some text"`), numbers, variables, mathematical expressions, etc. We'll learn more about variables & functions in the next tutorial.

Creating a code cell for each variable or mathematical operation can get tedious. Fortunately, Jupyter allows you to write multiple lines of code within a single code cell. The result of the last line of code within the cell is displayed as the output. 

Let's rewrite the solution to our word problem within a single cell.

In [29]:
# Store input data in variables
# this is a normal text that doesnt run
cost_of_ice_bag = 1.25
profit_margin = .5
number_of_bags = 500

# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

# Display the result
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 312.5


Note that we're using the `#` character to add *comments* within our code. 

> **Comments**: Comments and blank lines are ignored during execution, but they are useful for providing information to humans (including yourself) about what the code does. Comments can be inline (at the end of some code), on a separate line, or even span multiple lines. 

Inline and single-line comments start with `#`, whereas multi-line comments begin and end with three quotes, i.e. `"""`. Here are some examples of code comments:

In [31]:
my_favorite_number = 1 # favourite number input

In [32]:
# This comment gets its own line
my_least_favorite_number = 3

In [ ]:
"""This is a multi-line comment.
Write as little or as much as you'd like.

Comments are really helpful for people reading
your code, but try to keep them short & to-the-point.

Also, if you use good variable names, then your code is
often self explanatory, and you may not even need comments!
"""
a_neutral_number = 5



> **EXERCISE**: A travel company wants to fly a plane to the Bahamas. Flying the plane costs 5000 dollars. So far, 29 people have signed up for the trip. If the company charges 200 dollars per ticket, what is the profit made by the company? Create variables for each numeric quantity and use appropriate arithmetic operations.

## Evaluating conditions using Python

Apart from arithmetic operations, Python also provides several operations for comparing numbers & variables.

| Operator    | Description                                                     |
|-------------|-----------------------------------------------------------------|
| `==`        | Check if operands are equal                                     |
| `!=`        | Check if operands are not equal                                 |
| `>`         | Check if left operand is greater than right operand             |
| `<`         | Check if left operand is less than right operand                |
| `>=`        | Check if left operand is greater than or equal to right operand |
| `<=`        | Check if left operand is less than or equal to right operand    |

The result of a comparison operation is either `True` or `False` (note the uppercase `T` and `F`). These are special keywords in Python. Let's try out some experiments with comparison operators.

In [34]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [35]:
# Equality check - True
my_favorite_number == 1

True

In [36]:
# Equality check - False
my_favorite_number == my_least_favorite_number

False

In [37]:
# Not equal check - True
my_favorite_number != a_neutral_number

True

In [38]:
# Not equal check - False
a_neutral_number != 3

False

In [ ]:
# Greater than check - True
my_least_favorite_number > a_neutral_number

In [39]:
# Greater than check - False
my_favorite_number > my_least_favorite_number

False

In [40]:
# Less than check - True
my_favorite_number < 10

True

In [ ]:
# Less than check - False
my_least_favorite_number < my_favorite_number

In [ ]:
# Greater than or equal check - True
my_favorite_number >= 1

In [41]:
# Greater than or equal check - False
my_favorite_number >= 3

False

In [ ]:
# Less than or equal check - True
3 + 6 <= 9

In [42]:
# Less than or equal check - False
my_favorite_number + a_neutral_number <= 3
# 1+3 <=3

False

Just like arithmetic operations, the result of a comparison operation can also be stored in a variable.

In [45]:
cost_of_ice_bag = 1.25
is_ice_bag_expensive = cost_of_ice_bag >= 10
# print("Is the ice bag expensive?", is_ice_bag_expensive)
print ("this is from print statement")

this is from print statement


## Combining conditions with logical operators

The logical operators `and`, `or` and `not` operate upon conditions and `True` & `False` values (also known as *booleans*). `and` and `or` operate on two conditions, whereas `not` operates on a single condition.

The `and` operator returns `True` when both the conditions evaluate to `True`. Otherwise, it returns `False`.

| `a`     | `b`    | `a and b` |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |


In [46]:
my_favorite_number

1

In [47]:
my_favorite_number > 0 and my_favorite_number <= 3

True

In [48]:
my_favorite_number < 0 and my_favorite_number <= 3

False

In [ ]:
my_favorite_number > 0 and my_favorite_number >= 3

In [50]:
True and False

False

In [49]:
True and True

True

The `or` operator returns `True` if at least one of the conditions evaluates to `True`. It returns `False` only if both conditions are `False`.

| `a`     | `b`    | `a or b`  |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |



In [51]:
a_neutral_number = 3

In [52]:
a_neutral_number == 3 or my_favorite_number < 0

True

In [ ]:
a_neutral_number != 3 or my_favorite_number < 0

In [ ]:
my_favorite_number < 0 or True

In [54]:
False or False or False

False

The `not` operator returns `False` if a condition is `True` and `True` if the condition is `False`.

In [55]:
not a_neutral_number == 3

False

In [ ]:
not my_favorite_number < 0

In [56]:
not False

True

In [57]:
not True

False

Logical operators can be combined to form complex conditions. Use round brackets or parentheses `(` and `)` to indicate the order in which logical operators should be applied.

In [58]:
(2 > 3 and 4 <= 5) or not (my_favorite_number < 0 and True)

True

In [ ]:
not (True and 0 < 1) or (False and True)

If parentheses are not used, logical operators are applied from left to right.

In [ ]:
not True and 0 < 1 or False and True

# heading 1
## heading 2

**adasdad**

_italic_

Experiment with arithmetic, conditional and logical operators in Python using the interactive nature of Jupyter notebooks. We will learn more about variables and functions in future tutorials.

## Adding text styles using Markdown

Adding explanations using text cells (like this one) is a great way to make your notebook informative for other readers. It is also useful if you need to refer back to it in the future. You can double click on a text cell within Jupyter to edit it. In the edit mode, you'll notice that the text looks slightly different (for instance, the heading has a `##` prefix. This text is written using Markdown, a simple way to add styles to your text. Execute this cell to see the output without the special characters. You can switch back and forth between the source and the output to apply a specific style.

For instance, you can use one or more `#` characters at the start of a line to create headers of different sizes:

# Header 1

## Header 2

### Header 3

#### Header 4

To create a bulleted or numbered list, simply start a line with `*` or `1.`.

A bulleted list:

* Item 1
* Item 2
* Item 3

A check Boxes

- [ ] sadasd
- [x] dsada


A numbered list:

1. Apple
2. Banana
3. Pineapple
4. Nothing

You can make some text bold using `**`, e.g., **some bold text**, or make it italic using `*`, e.g., *some italic text.* You can also create links, e.g., [GIthub](https://github.com). Images are easily embedded too:

![](https://i.imgur.com/3gjZMYK.png)

Another really nice feature of Markdown is the ability to include blocks of code. Note that code blocks inside Markdown cells cannot be executed.

```
# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

# Display the result
print("The grocery store makes a total profit of $", total_profit)

```

You can learn the full syntax of Markdown here: https://learnxinyminutes.com/docs/markdown/

## Further Reading and References

Following are some resources where you can learn about more arithmetic, conditional and logical operations in Python:

* Python Tutorial at W3Schools: https://www.w3schools.com/python/
* Practical Python Programming: https://dabeaz-course.github.io/practical-python/Notes/Contents.html
* Python official documentation: https://docs.python.org/3/tutorial/index.html

Now that you have taken your first steps with Python, you are ready to move on to the next tutorial: ["Day 1 Part 2"](https://drive.google.com/file/d/11YYPapz_OAqR9TuZudj-g4H-OlynQf0P/view?usp=sharing)

## Questions for Revision

Try answering the following questions to test your understanding of the topics covered in this notebook:

1. What is a Jupyter notebook? 
2. How do you add a new code cell below an existing cell?
3. How do you add a new Markdown cell below an existing cell?
4. How do you convert a code cell to a Markdown cell or vice versa?
5. How do you execute a code cell within Jupyter?
6. What the different arithmetic operations supported in Python?
7. How do you perform arithmetic operations using Python?
8. What is the difference between the `/` and the `//` operators?
9. What is the difference between the `*` and the `**` operators?
10. What is the order of precedence for arithmetic operators in Python?
11. How do you specify the order in which arithmetic operations are performed in an expression involving multiple operators?
12. How do you solve a multi-step arithmetic word problem using Python?
13. What are variables? Why are they useful?
14. How do you create a variable in Python?
15. What is the assignment operator in Python?
16. What are the rules for naming a variable in Python?
17. How do you view the value of a variable?
18. How do you store the result of an arithmetic expression in a variable?
19. What happens if you try to access a variable that has not been defined?
20. How do you display messages in Python?
21. What type of inputs can the print function accept?
22. What are code comments? How are they useful?
23. What are the different ways of creating comments in Python code?
24. What are the different comparison operations supported in Python?
25. What is the result of a comparison operation?
26. What is the difference between `=` and `==` in Python?
27. What are the logical operators supported in Python?
28. What is the difference between the `and` and `or` operators?
29. Can you use comparison and logical operators in the same expression?
30. What is the purpose of using parentheses in arithmetic or logical expressions?
31. What is Markdown? Why is it useful?
32. How do you create headings of different sizes using Markdown?
33. How do you create bulleted and numbered lists using Markdown?
34. How do you create bold or italic text using Markdown?
35. How do you include links & images within Markdown cells?
36. How do you include code blocks within Markdown cells?
37. Is it possible to execute the code blocks within Markdown cells?
38. How do you upload and share your Jupyter notebook online using Jovian?
39. What is the purpose of the API key requested by jovian.commit ? Where can you find the API key?
40. Where can you learn about arithmetic, conditional and logical operations in Python?
